ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs

```diff
@@ -496,6 +529,44 @@ chainSyncClient tracer cfg btime (ClockSkew maxSkew)
     k :: Word64
     k = maxRollbacks $ protocolSecurityParam cfg
 
+-- | Watch the invalid block checker function for changes (using its
+-- fingerprint). Whenever it changes, i.e., a new invalid block is detected,
+-- check whether the current candidate fragment contains any header that is
+-- invalid, if so, disconnect by throwing an 'InvalidBlock' exception.
+--
+-- This function spawns a background thread using the given 'ThreadRegistry'.
+--
+-- The cost of this check is \( O(cand * check) \) where /cand/ is the size of
+-- the candidate fragment and /check/ is the cost of checking whether a block
+-- is invalid (typically \( O(\log(invalid)) \) where /invalid/ is the number
+-- of invalid blocks).
```

Comment:
Might be worth pointing out here that in the possible though unlikely scenario that the node gives us an invalid block, then rolls back and gives us a valid block, we might not disconnect.